arch/arm/core/aarch32/cpu_idle.S

```diff
@@ -44,86 +44,119 @@ GTEXT(arch_cpu_atomic_idle)
 
 SECTION_FUNC(TEXT, z_arm_cpu_idle_init)
 #if defined(CONFIG_CPU_CORTEX_M)
-	ldr r1, =_SCB_SCR
-	movs.n r2, #_SCR_INIT_BITS
-	str r2, [r1]
+	ldr	r1, =_SCB_SCR
+	movs.n	r2, #_SCR_INIT_BITS
+	str	r2, [r1]
 #endif
-	bx lr
+	bx	lr
 
 SECTION_FUNC(TEXT, arch_cpu_idle)
 #ifdef CONFIG_TRACING
-	push {r0, lr}
-	bl    sys_trace_idle
+	push	{r0, lr}
+	bl	sys_trace_idle
 #if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
-	pop {r0, r1}
-        mov lr, r1
+	pop	{r0, r1}
+	mov	lr, r1
 #else
-	pop {r0, lr}
+	pop	{r0, lr}
 #endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
 #endif /* CONFIG_TRACING */
 
-#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE) \
-	|| defined(CONFIG_ARMV7_R)
-	cpsie i
-#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
-	/* clear BASEPRI so wfi is awakened by incoming interrupts */
-	eors.n r0, r0
-	msr BASEPRI, r0
+#if defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
+	/*
+	 * PRIMASK is always cleared on ARMv7-M and ARMv8-M Mainline, and
+	 * configuring BASEPRI to the lowest priority to ensure wake-up will
+	 * cause interrupts to be serviced before entering low power state.
+	 *
+	 * Set PRIMASK before configuring BASEPRI to prevent interruption
+	 * before wake-up.
+	 */
+	cpsid	i
+
+	/*
+	 * Set wake-up interrupt priority to the lowest and synchronise to
+	 * ensure that this is visible to the WFI instruction.
+	 */
+	eors.n	r0, r0
+	msr	BASEPRI, r0
+	isb
 #else
-#error Unknown ARM architecture
-#endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
+	/*
+	 * For all the other ARM architectures that do not implement BASEPRI,
+	 * PRIMASK is used as the interrupt locking mechanism, and it is not
+	 * necessary to set PRIMASK here, as PRIMASK would have already been
+	 * set by the caller as part of interrupt locking if necessary
+	 * (i.e. if the caller sets _kernel.idle).
+	 */
+#endif /* CONFIG_ARMV7_M_ARMV8_M_MAINLINE */
+
+	/*
+	 * Wait for all memory transactions to complete before entering low
```

Comment:
Yeap, descriptions is now much better.
Although an un-experience ARM developer may still confuse this with DMB.
Strictly speaking DSB means "No instruction in program order after this instruction can execute until this instruction completes."
But I am OK with the current text.